File: nominas.cpp

// MTI Antonio Armando Presuel Rosado
//
// NOMINA POR MEDIO DE ARCHIVO DE ACCESO DIRECTO.
//

#include <iostream>
#include <fstream>
#include <cstdlib> // system()
#include <iomanip> // para setw() y system()
#include "docente.h"
using namespace std;

int main(){
    // ------- Constantes empleadas
    const int sizeCampo1= 30;
    const int sizeCampo2= 30;
    const int sizeCampo3= 6; 
    const int posCampo2= sizeCampo1;
    const int posCampo3= posCampo2+sizeCampo2;
    const int posCampo4= posCampo3+sizeCampo3;
    const int posCampo5= posCampo4+sizeof(int);
    const int sizeReg= sizeCampo1+sizeCampo2+sizeCampo3+sizeof(int)+sizeof(double);
    const int base= 71;
    // ------- Archivos
    string nombreArchivo= "a_plantilla.txt";
    string ArchivoNomina= "b_nomina.txt";
    fstream archivo;    
    archivo.open(nombreArchivo, ios::in);
    if(!archivo){
        archivo.open(nombreArchivo, ios::out);
        archivo << "AAPR20220516ITS1412101NATECNOLOGICO NACIONAL DE MEXICO CAMPUS HOPELCHEN";
    }
    archivo.close();
    // ------- Iteración del menú
    int opcion;
    do{
        system("cls");
        // ------- Menú
        cout << "Que accion desea realizar?\n";
        cout << "  1- Generar nomina.\n";
        cout << "  2- Ver lista de empleados.\n";
        cout << "  3- Agregar empleado.\n";
        cout << "  4- Consultar empleado.\n";
        cout << "  0- Salir del programa.\n";
        cout << "Opcion: ";
        cin >> opcion;
        cin.ignore();
        //
        switch(opcion){
            case 1:{ /////////////////////////////////////////////////////////// GENERAR NOMINA
                // ------- Datos para la actividad
                int registros;
                string autor= "", rfc= "", empresa= "";
                double montoTotal= 0;
                string cadena;
                // ------- Abriendo el archivo
                fstream nomina;
                nomina.open(ArchivoNomina, ios::out);
                archivo.open(nombreArchivo, ios::in);
                archivo.seekg (0, ios::beg);
                // ------- Encabezado
                for(int i=0; i<base; i++){
                    char x= archivo.get();
                    (i<12)? autor += x : (i<24)? rfc += x : empresa += x;
                }
                system("cls");
                cadena= "\n"+empresa+" ("+rfc+")\n\n";
                nomina.write(cadena.c_str(), cadena.length());
                cadena= "Nomina de la plantilla docente.\n";
                nomina.write(cadena.c_str(), cadena.length());
                nomina << setw( 76 ) << setfill( '-' ) << '\n' << setfill( ' ' );
                nomina << "| " << right << setw( 8 ) << "Clave  " << "|"
                       << left << setw( 50 ) << " Empleado" << " | "
                       << right << setw( 10 ) << "Sueldo  " << "|\n";
                nomina << setw( 76 ) << setfill( '-' ) << '\n' << setfill( ' ' );
                // ------- Calculando registros en plantilla
                archivo.seekg (0, archivo.end);
                registros = archivo.tellg();
                registros= (registros-base)/sizeReg;
                // ------- Recuperando y presentando los datos
                for(int i=0; i<registros; i++){
                    // ------- Datos para la actividad
                    string nombres, apellidos, clave;
                    double pagoHora;
                    int horasAsignadas;
                    int baseRegistro= base+(sizeReg*i);
                    char x;
                    // ---
                    archivo.seekg (baseRegistro, archivo.beg);                    
                    while( (x= archivo.get()) && x != 30 ){ nombres += x; }
                    archivo.seekg (baseRegistro+posCampo2, archivo.beg);
                    while( (x= archivo.get()) && x != 30 ){ apellidos += x; } 
                    archivo.seekg (baseRegistro+posCampo3, archivo.beg);
                    while( (x= archivo.get()) && x != 30 ){ clave += x; } 
                    archivo.seekg (baseRegistro+posCampo4, archivo.beg);                    
                    archivo.read((char *) &horasAsignadas, sizeof(int));
                    archivo.seekg (baseRegistro+posCampo5, archivo.beg);
                    archivo.read((char *) &pagoHora, sizeof(double));
                    // ---
                    double subMonto= int(horasAsignadas)*double(pagoHora)*2;
                    montoTotal+= subMonto;
                    nomina << "| " << right << setw( 7 ) << clave << " | "
                           << left << setw( 49 ) << (apellidos + ", " + nombres) << " | "
                           << right << setw( 9 ) << fixed << setprecision(2) << subMonto << " |" << endl;
                }
                nomina << setw( 76 ) << setfill( '-' ) << '\n' << setfill( ' ' );
                nomina << "Monto de la nomina: $";
                nomina << fixed << setprecision(2) << montoTotal << " pesos.\n";
                nomina.close();
                archivo.close();
                cout << "El monto de la nomina es de $" << montoTotal << " pesos.\n";
                cout << "Se ha generado la nomina en el archivo correspondiente.\n";
                system("pause");
            }
            break;
            case 2:{ ////////////////////////////////////////////////////////// LISTAR PLANTILLA
                // ------- Datos para la actividad
                int registros;
                string autor= "", rfc= "", empresa= "";
                // ------- Abriendo el archivo
                archivo.open(nombreArchivo, ios::in);                
                archivo.seekg (0, ios::beg);
                // Encabezado
                for(int i=0; i<base; i++){
                    char x= archivo.get();
                    (i<12)? autor += x : (i<24)? rfc += x : empresa += x;
                }
                system("cls");
                cout << "\n" << empresa << " (" << rfc << ")\n" << ".:. " << autor << " .:.\n\n";
                cout << "Plantilla docente.\n------------------------------------------------\n";
                // ------- Calculando registros en plantilla
                archivo.seekg (0, archivo.end);
                registros = archivo.tellg();
                registros= (registros-base)/sizeReg;
                // ------- Recuperando y presentando los datos
                for(int i=0; i<registros; i++){
                    // ------- Datos para la actividad
                    string nombres, apellidos, clave;
                    double pagoHora;
                    int horasAsignadas;
                    int baseRegistro= base+(sizeReg*i);
                    char x;
                    // ---
                    archivo.seekg (baseRegistro, archivo.beg);                    
                    while( (x= archivo.get()) && x != 30 ){ nombres += x; }
                    archivo.seekg (baseRegistro+posCampo2, archivo.beg);
                    while( (x= archivo.get()) && x != 30 ){ apellidos += x; } 
                    archivo.seekg (baseRegistro+posCampo3, archivo.beg);
                    while( (x= archivo.get()) && x != 30 ){ clave += x; } 
                    archivo.seekg (baseRegistro+posCampo4, archivo.beg);                    
                    archivo.read((char *) &horasAsignadas, sizeof(int));
                    archivo.seekg (baseRegistro+posCampo5, archivo.beg);
                    archivo.read((char *) &pagoHora, sizeof(double));
                    // ---
                    cout << i+1 << ") " << clave << " - " 
                         << apellidos << ", " << nombres << ". Horas: " 
                         << horasAsignadas << ", $" << pagoHora << " pesos.\n";
                }
                archivo.close();
                cout << "\n";
                system("pause");
            }
            break;
            case 3:{ /////////////////////////////////////////////////////// AGREGAR EMPLEADO                
                // ------- Datos para la actividad
                docente empleado;
                string cadena;
                int entero;
                double real;
                // ------- Solicitando datos al usuario
                cout << "Nombres: ";
                getline(cin, cadena);
                empleado.setNombres(cadena);
                cout << "Apellidos: ";
                getline(cin, cadena);
                empleado.setApellidos(cadena);
                cout << "Clave: ";
                getline(cin, cadena);
                empleado.setClave(cadena);
                cout << "Pago por hora: ";
                cin >> real; 
                empleado.setPagoHora(real);
                cout << "Horas asignadas: ";
                cin >> entero;
                empleado.setHorasContrato(entero);
                cin.ignore();
                // --------- Apertura de archivo, apuntando al final.
                archivo.open(nombreArchivo, ios::app);
                archivo.seekp (0, ios::end);
                // --------- Almacenamiento en archivo
                cadena= empleado.getNombres() + (char)30;
                archivo.write(cadena.c_str(), sizeCampo1);
                cadena= empleado.getApellidos() + (char)30;
                archivo.write(cadena.c_str(), sizeCampo2);
                cadena= empleado.getClave()+(char)30;
                archivo.write(cadena.c_str(), sizeCampo3);                
                entero= empleado.getHorasContrato();
                archivo.write((char *)&entero, sizeof(int));
                real= empleado.getPagoHora();
                archivo.write((char *)&real, sizeof(double));
                // --------- Finalizando         
                archivo.close();
                cout << "Registro agregado!\n";
                system("pause");
                system("cls"); 
            }
            break;
            case 4:{ //////////////////////////////////////////////////////////// CONSULTAR EMPLEADO                
                // ------- Datos para la actividad
                int registros, baseRegistro, registroBuscado, horasAsignadas;                
                string nombres, apellidos, clave;
                double pagoHora;                
                char x;
                // ------- Calculando registros en plantilla
                archivo.open(nombreArchivo, ios::in);
                archivo.seekg (0, archivo.end);
                registros = archivo.tellg();
                registros= (registros-base)/sizeReg;
                cout << "\n-------\nSe encontraron " << registros << " registros.\n"; 
                cout << "Ingrese el numero de registro a consultar: ";                
                cin >> registroBuscado; // pendiente validar con if | try.
                (registroBuscado>0)? registroBuscado-- : registroBuscado= 0;
                baseRegistro= base+(sizeReg*registroBuscado);
                // ------- Recuperando y presentan el registro deseado         
                archivo.seekg (baseRegistro, archivo.beg);
                while( (x= archivo.get()) && x != 30 ){ nombres += x; }
                archivo.seekg (baseRegistro+posCampo2, archivo.beg);
                while( (x= archivo.get()) && x != 30 ){ apellidos += x; } 
                archivo.seekg (baseRegistro+posCampo3, archivo.beg);
                while( (x= archivo.get()) && x != 30 ){ clave += x; } 
                archivo.seekg (baseRegistro+posCampo4, archivo.beg);                    
                archivo.read((char *) &horasAsignadas, sizeof(int));
                archivo.seekg (baseRegistro+posCampo5, archivo.beg);
                archivo.read((char *) &pagoHora, sizeof(double));       
                // ------- Presentación del dato recuperado
                cout << clave << " - " << apellidos << ", " << nombres << ". Horas: " 
                        << horasAsignadas << ", $" << pagoHora << " pesos.\n";
                archivo.close();
                cout << "\n";
                system("pause");
                system("cls");
            }
            break;
            case 0:{ // FINALIZA EL PROGRAMA
                cout << "Finalizando el programa!\n";
            }
            break;
        }
    }while(opcion != 0);
    //
    system("pause");
    return 0;
}

File: nominas.java

// MTI Antonio Armando Presuel Rosado
//
// NOMINA POR MEDIO DE ARCHIVO DE ACCESO DIRECTO.
//
// Long     8 bytes     -   Integer     4 bytes     -   Short       2 bytes
// Byte     1 byte      -   Double      8 bytes     -   Float       4 bytes
// Boolean  1 byte      -   Char        2 bytes     -   String      2 bytes por cada caracter.
// 
// Ojo: al posicionar, en el archivo se recorren los strings x2 y en el programa al contar se avanza x1.
//
import java.io.FileWriter;
import java.io.RandomAccessFile;
import java.util.Scanner;

public class nominas {

    public static String removerCaracteresEspeciales(String s) {
        if (s == null)
            return null;
        return s.replaceAll("[^\\sA-Za-z0-9]", "");
    }

    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);
        System.out.println("\nBienvenido al sistema de calculo de nomina.\n---");
        // constantes empleadas
        String encabezado = "AAPR20220611ITS1412101NATECNOLOGICO NACIONAL DE MEXICO CAMPUS HOPELCHEN";
        int base = encabezado.length() * 2;
        int sizeCampo1 = 30; // nombres
        int sizeCampo2 = 30; // apellidos
        int sizeCampo3 = 6; // clave
        int posCampo1 = 0;
        int posCampo2 = posCampo1 + sizeCampo1 * 2;
        int posCampo3 = posCampo2 + sizeCampo2 * 2;
        int posCampo4 = posCampo3 + sizeCampo3 * 2;
        int posCampo5 = posCampo4 + 4;
        int sizeRegFile = posCampo5 + 8;
        // archivos
        String nombreArchivo = "aJv_plantilla.txt";
        RandomAccessFile archivo;
        try {
            archivo = new RandomAccessFile(nombreArchivo, "rw");
            if (archivo.length() == 0) {
                archivo.seek(0);
                StringBuffer bufer;
                bufer = new StringBuffer(encabezado);
                bufer.setLength(encabezado.length());
                archivo.writeChars(bufer.toString());
            } else {
                archivo.seek(0);
                String firma = "", rfc = "", institucion = "";
                for (int i = 0; i < 71; i++) {
                    if (i < 12)
                        firma += archivo.readChar();
                    else if (i < 24)
                        rfc += archivo.readChar();
                    else
                        institucion += archivo.readChar();
                }
                System.out.println("Firma: " + firma + "\nRFC: " + rfc + "\nInstitucion: " + institucion + "\n---\n");
            }
            archivo.close();
        } catch (Exception e) {
            e.printStackTrace();
        }
        // iteracion del menú
        int opcion;
        do {
            System.out.println("Que accion desea realizar?");
            System.out.println("  1- Generar nomina.");
            System.out.println("  2- Ver lista de empleados.");
            System.out.println("  3- Agregar empleado.");
            System.out.println("  4- Consultar empleado.");
            System.out.println("  0- Salir del programa.");
            System.out.print("Opcion: ");
            opcion = scan.nextInt();
            scan.nextLine();
            switch (opcion) {
                case 1: {/////////////////////////////////////////////////////////// GENERAR NOMINA
                    // ------- Datos para la actividad
                    String ArchivoNomina = "bJv_nomina.txt";
                    FileWriter nomina;
                    int registros, baseRegistro;
                    String firma = "", rfc = "", institucion = "";
                    double montoTotal = 0;
                    String cadena;
                    docente empleado = new docente();
                    // -------
                    try {
                        // ------- Abriendo el archivo
                        archivo = new RandomAccessFile(nombreArchivo, "rw");
                        nomina = new FileWriter(ArchivoNomina, true);
                        // Encabezado
                        archivo.seek(0);
                        for (int i = 0; i < 71; i++) {
                            if (i < 12)
                                firma += archivo.readChar();
                            else if (i < 24)
                                rfc += archivo.readChar();
                            else
                                institucion += archivo.readChar();
                        }
                        // ------- Calculando registros en plantilla
                        registros = (int) archivo.length();
                        registros = (registros - base) / sizeRegFile;
                        // Impresion en pantalla
                        cadena = "\nSe genera nomina para " + registros + " empleados.";
                        System.out.print(cadena);
                        // Impresion en archivo de Texto
                        cadena = "\n" + institucion + " (" + rfc + ")\n\n";
                        nomina.write(cadena);
                        cadena = "Nomina de la plantilla docente.\n";
                        nomina.write(cadena);
                        cadena = "-";
                        for (int i = 0; i < 71; i++)
                            nomina.write(cadena);
                        cadena = String.format("\n%-8s| %-50s| %-10s|\n", "Clave", "Empleado", "Sueldo");
                        nomina.write(cadena);
                        cadena = "-";
                        for (int i = 0; i < 71; i++)
                            nomina.write(cadena);
                        // ------- Recuperando y presentando los datos
                        for (int r = 0; r < registros; r++) {
                            baseRegistro = base + (sizeRegFile * r);
                            // ------- Recuperando el registro
                            archivo.seek(baseRegistro + posCampo1);//
                            cadena = "";
                            for (int i = 0; i < sizeCampo1; i++)
                                cadena += archivo.readChar();
                            empleado.setNombres(removerCaracteresEspeciales(cadena));
                            archivo.seek(baseRegistro + posCampo2);//
                            cadena = "";
                            for (int i = 0; i < sizeCampo2; i++)
                                cadena += archivo.readChar();
                            empleado.setApellidos(removerCaracteresEspeciales(cadena));
                            archivo.seek(baseRegistro + posCampo3);//
                            cadena = "";
                            for (int i = 0; i < sizeCampo3; i++)
                                cadena += archivo.readChar();
                            empleado.setClave(removerCaracteresEspeciales(cadena));
                            archivo.seek(baseRegistro + posCampo4);//
                            empleado.setHorasContrato(archivo.readInt());
                            archivo.seek(baseRegistro + posCampo5);//
                            empleado.setPagoHora(archivo.readDouble());
                            // Presentando los datos recuperados
                            String nombreCompleto = empleado.getApellidos() + " " + empleado.getNombres();
                            double Sueldo = empleado.getHorasContrato() * empleado.getPagoHora() * 2;
                            montoTotal += Sueldo;
                            cadena = String.format("\n%-8s| %-50s| %-10s|\n", empleado.getClave(),
                                    nombreCompleto, Sueldo);
                            nomina.write(cadena);
                        }
                        cadena = "-";
                        for (int i = 0; i < 71; i++)
                            nomina.write(cadena);
                        cadena = "\nMonto de la nomina: " + montoTotal + "\n";
                        nomina.write(cadena);
                        System.out.println(cadena);
                        cadena = ("((( " + firma + " )))\n");
                        nomina.write(cadena);
                        // -------
                        nomina.close();
                        archivo.close();
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                }
                    break;
                case 2: {////////////////////////////////////////////////////////// LISTAR PLANTILLA
                    // ------- Datos para la actividad
                    String cadena;
                    docente empleado = new docente();
                    String firma = "", rfc = "", institucion = "";
                    int registros, baseRegistro;
                    // --------- Apertura de archivo para lectura/escritura
                    try {
                        // Encabezado
                        archivo = new RandomAccessFile(nombreArchivo, "rw");
                        archivo.seek(0);
                        for (int i = 0; i < 71; i++) {
                            if (i < 12)
                                firma += archivo.readChar();
                            else if (i < 24)
                                rfc += archivo.readChar();
                            else
                                institucion += archivo.readChar();
                        }
                        System.out.println(
                                "Firma: " + firma + "\nRFC: " + rfc + "\nInstitucion: " + institucion);
                        // ------- Calculando cantidad de registros
                        registros = (int) archivo.length();
                        registros = (registros - base) / sizeRegFile;
                        //
                        for (int r = 0; r < registros; r++) {
                            baseRegistro = base + (sizeRegFile * r);
                            // ------- Recuperando el registro
                            archivo.seek(baseRegistro + posCampo1);//
                            cadena = "";
                            for (int i = 0; i < sizeCampo1; i++)
                                cadena += archivo.readChar();
                            empleado.setNombres(cadena);
                            archivo.seek(baseRegistro + posCampo2);//
                            cadena = "";
                            for (int i = 0; i < sizeCampo2; i++)
                                cadena += archivo.readChar();
                            empleado.setApellidos(cadena);
                            archivo.seek(baseRegistro + posCampo3);//
                            cadena = "";
                            for (int i = 0; i < sizeCampo3; i++)
                                cadena += archivo.readChar();
                            empleado.setClave(cadena);
                            archivo.seek(baseRegistro + posCampo4);//
                            empleado.setHorasContrato(archivo.readInt());
                            archivo.seek(baseRegistro + posCampo5);//
                            empleado.setPagoHora(archivo.readDouble());
                            // Presentando los datos recuperados
                            System.out.print("------- -------\n" + empleado.getClave() + " | ");
                            System.out.println(empleado.getApellidos() + " | " + empleado.getNombres());
                            System.out.println(empleado.getHorasContrato() + " horas | $ " + empleado.getPagoHora());
                            System.out.println("------- -------");
                        }
                        // --- finalizando
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                }
                    break;
                case 3: {/////////////////////////////////////////////////////// AGREGAR EMPLEADO
                    docente empleado = new docente();
                    // ------- Solicitando datos al usuario
                    System.out.println("Nombres: ");
                    empleado.setNombres(scan.nextLine());
                    System.out.println("Apellidos: ");
                    empleado.setApellidos(scan.nextLine());
                    System.out.println("Clave: ");
                    empleado.setClave(scan.nextLine());
                    System.out.println("Pago por hora: ");
                    empleado.setPagoHora(scan.nextDouble());
                    System.out.println("Horas asignadas: ");
                    empleado.setHorasContrato(scan.nextInt());
                    scan.nextLine();
                    // --------- Apertura de archivo para lectura/escritura
                    try {
                        // --------- Almacenamiento en archivo
                        archivo = new RandomAccessFile(nombreArchivo, "rw");
                        archivo.seek(archivo.length());
                        StringBuffer bufer;
                        bufer = new StringBuffer(empleado.getNombres()); //
                        bufer.setLength(sizeCampo1);
                        archivo.writeChars(bufer.toString());
                        bufer = new StringBuffer(empleado.getApellidos()); //
                        bufer.setLength(sizeCampo2);
                        archivo.writeChars(bufer.toString());
                        bufer = new StringBuffer(empleado.getClave()); //
                        bufer.setLength(sizeCampo3);
                        archivo.writeChars(bufer.toString());
                        archivo.writeInt(empleado.getHorasContrato()); //
                        archivo.writeDouble(empleado.getPagoHora()); //
                        // --------- Finalizando
                        archivo.close();
                        System.out.println("Registro agregado.");
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                }
                    break;
                case 4: {//////////////////////////////////////////////////////////// CONSULTAR EMPLEADO
                    // ------- Datos para la actividad
                    docente empleado = new docente();
                    int registros, baseRegistro, registroBuscado;
                    String cadena;
                    // --------- Apertura de archivo para lectura
                    try {
                        archivo = new RandomAccessFile(nombreArchivo, "r");
                        // ------- Calculando registros en plantilla
                        registros = (int) archivo.length();
                        registros = (registros - base) / sizeRegFile;
                        System.out.println("Se encontraron " + registros + " registros.");
                        // ------- Ubicando el registro buscado
                        System.out.println("Ingrese el numero de registro a consultar: ");
                        registroBuscado = scan.nextInt();
                        registroBuscado = (registroBuscado > 0) ? registroBuscado - 1 : 0;
                        baseRegistro = base + (sizeRegFile * registroBuscado);
                        // ------- Recuperando el registro deseado
                        archivo.seek(baseRegistro + posCampo1);//
                        cadena = "";
                        for (int i = 0; i < sizeCampo1; i++) {
                            cadena += archivo.readChar();
                        }
                        empleado.setNombres(cadena);
                        archivo.seek(baseRegistro + posCampo2);//
                        cadena = "";
                        for (int i = 0; i < sizeCampo2; i++) {
                            cadena += archivo.readChar();
                        }
                        empleado.setApellidos(cadena);
                        archivo.seek(baseRegistro + posCampo3);//
                        cadena = "";
                        for (int i = 0; i < sizeCampo3; i++) {
                            cadena += archivo.readChar();
                        }
                        empleado.setClave(cadena);
                        archivo.seek(baseRegistro + posCampo4);//
                        empleado.setHorasContrato(archivo.readInt());
                        archivo.seek(baseRegistro + posCampo5);//
                        empleado.setPagoHora(archivo.readDouble());
                        // Finalizando
                        archivo.close();
                        System.out.print("------- -------\n" + empleado.getClave() + " | ");
                        System.out.println(empleado.getApellidos() + " | " + empleado.getNombres());
                        System.out.println(empleado.getHorasContrato() + " horas | $ " + empleado.getPagoHora());
                        System.out.println("------- -------");
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                }
                    break;
                case 0:
                    System.out.println("Finalizando el programa.");
                    break;
            }
        } while (opcion != 0);
        // Finalización
        scan.close();
    }
}